frontend/model/captureLogs.js

```diff
@@ -0,0 +1,250 @@
+  // entries limit is changed. Example: There are 400 logs and the limit is 500.
+  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+  // We do it recursively in chunks of 25 until there's only 75 logs again.
+  while (entriesCount >= ENTRIES_LIMIT && failedDeletionCount < 3) {
+    markers = markers || getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+
+    if (entriesCount === lastEntriesCount) { failedDeletionCount++ }
+    lastEntriesCount = entriesCount
+
+    do {
+      const log = JSON.parse(giLSget(oldestEntry)) || {}
+      giLSremove(oldestEntry)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+  }
+
+  if (failedDeletionCount < 3) {
+    giLSset('markers', JSON.stringify(markers))
+    giLSset('count', entriesCount)
+  } else {
+    // - There are too many entriesCount, however, for some unknown reason
+    //   we weren't able to remove them.
+    // - Capture a new log error manually (withou console.error) so
+    //   that verifyLogsSize() doesn't get called again. There's no need
+    //   for that and it would cause a loop of failed deletions.
+    captureLog('error', 'verifyLogsSize() failed! Unable to delete oldest logs.')
+
+    // To avoid infinite loops or memory leaks, let's just pause the capturing.
+    captureLogsPause()
+    // Or... if we want to be extreme, clear the logs and have a fresh start?
+    // clearLogs()
+  }
+}
+
+function verifyLogsConfigs () {
+  // If ENTRIES_LIMIT or ENTRIES_MARKER_NTH are changed in a release
+  // we need to recalculate markers and verify if it reached the size limit.
+  const storedLimit = +giLSget('limit')
+  const storedMarkerNth = +giLSget('markerNth')
+
+  // when it's the first time running this, just set the correct limits
+  if (storedLimit === 0) {
+    giLSset('limit', ENTRIES_LIMIT)
+    giLSset('markerNth', ENTRIES_MARKER_NTH)
+    return
+  }
+
+  if (storedMarkerNth !== ENTRIES_MARKER_NTH) {
+    // recalculate markers based on new nth rule.
+    const newMarkers = []
+    let entryIndex = entriesCount
+    let prevEntry = lastEntry
+    do {
+      const log = JSON.parse(giLSget(`${prevEntry}`))
+      if (!log) break
+      if (entryIndex % ENTRIES_MARKER_NTH === 0) {
+        newMarkers.push(prevEntry)
+      }
+      entryIndex--
+      prevEntry = log.prev
+    } while (prevEntry)
+    // reverse new markers to be in chronological order, and then store them.
+    giLSset('markers', JSON.stringify(newMarkers.reverse()))
+    giLSset('markerNth', ENTRIES_MARKER_NTH)
+  }
+
+  if (storedLimit !== ENTRIES_LIMIT) {
+    giLSset('limit', ENTRIES_LIMIT)
+    verifyLogsSize()
+  }
+}
+
+export function captureLogsStart (userLogged) {
+  isCapturing = true
+  username = userLogged
+  lastEntry = giLSget('lastEntry')
+  entriesCount = +giLSget('count') || 0
```

Comment:
The code you mentioned doesn't work. In that while loop, we would need to `giLSget(curEntry)` for each while to access the .prev entry key. Do you really wanna do it when there are 4000 logs stored?

Please, tell me exactly the steps to get `entriesCount` out of sync. If it's a bug, it should be fixed. If the user/developer accidentally modified the localStorage, counting the logs wouldn't fix it because the modification could break the chain. Imagine we can only count 500 logs but the `localStorage.length` is 900. "_Something wrong is not right._" The chain was broken and in that case, we should just clear the logs and start over.

Reply:
> Please, tell me exactly the steps to get entriesCount out of sync. If it's a bug, it should be fixed. If the user/developer accidentally modified the localStorage, counting the logs could be useless because a possible "accidental modification" could break the chain.

I've seen it happen during testing, but I don't know how it happened. :(

> Imagine that we count 500 logs but the `localStorage.length` says there are 900 logs. "Something wrong is not right." The chain was broken and in that case, we should just clear the logs and start over.

`localStorage.length` is not a measure of the number of console logs, it's a measure of everything inside of localStorage, and we store quite a lot in there, so I'm not sure what you mean by this.

> And once again... the code you mentioned doesn't work. (it's incomplete). We would need to do `giLSget(curEntry)` inside for each while iteration to access the previous entry. 

Sorry Sandrina you are right 😭 , I did say that I didn't test the code and to please check it for errors! 😭

> Do you really want to do it, even when there are 4000 logs stored? I think it's a very expensive verification O(n) and it may not guarantee 100% effective results.

Heh, OK, if you implement clearing the logs and resetting everything on logout, then it doesn't need to be done, because if it gets out of sync for some reason, it's not the end of the world.

Reply:
At the moment we aren't storying almost anything at LS besides the `giConsole`. Even if we are, we can just look for the length of `giConsole/${username}` logs and compare it to `entriesCount`. If it doesn't match, it means there's something wrong.

> I don't know how it happened.

Okay.. then one tip: From now on, be self-aware of what steps exactly you do before encountering the bug and then let me know. And remember to reset the LS before you try next time to avoid old code from previous commits :)